src/loki/trace_route_action.cc

```diff
@@ -171,12 +171,78 @@ void loki_worker_t::locations_from_shape(valhalla_request_t& request) {
 
   // Add first and last correlated locations to request
   try {
+    // If trace route has 2 locations get the lat,lng of the first and last so we can support
+    // side of street.
+    bool has_locations = false;
+    PointLL orig_ll, dest_ll;
+    if (request.options.locations_size() == 2) {
+      has_locations = true;
+      orig_ll = {request.options.locations(0).ll().lng(), request.options.locations(0).ll().lat()};
+      dest_ll = {request.options.locations(1).ll().lng(), request.options.locations(1).ll().lat()};
+    }
+
+    // Project first and last shape point onto nearest edge(s). Clear current locations list
+    // and set the path locations
     auto projections = loki::Search(locations, *reader, edge_filter, node_filter);
     request.options.clear_locations();
     PathLocation::toPBF(projections.at(locations.front()), request.options.mutable_locations()->Add(),
                         *reader);
     PathLocation::toPBF(projections.at(locations.back()), request.options.mutable_locations()->Add(),
                         *reader);
+
+    // If locations were provided, backfill the origin and dest lat,lon and update
```

Comment:
why don't the `projections` above already have this information? why would we need to recompute it?

Reply:
if this is the trace from a prior route then the first and last points are on the edge so no side of street is provided. This allows an origin and a destination location to be passed in for these cases so a side of street can be determined. Code above basically clears out locations and uses the first and last shape point as inputs to Search.

Reply:
could we expose this function: https://github.com/valhalla/valhalla/blob/master/src/loki/search.cc#L89-L104

or something and call it from both search.cc and here so that we are both not duplicated the logic and making sure we have consistent behavior? maybe we could even move it to a util function in `baldr`?

Reply:
@dnesbitt61 were you going to try to reuse existing search logic?

Reply:
this is a member of candidate_t - which I don't think is readily available from the PathLocations or pbf. Also, the "duplicated code" is setting the pbf values directly, rather than the PathLocation members, so new code would still be needed to put these in the pbf.

Reply:
i was suggesting pulling that method out of `candidate_t` and just making it a stand along thing for computing the side of street given some inputs. this would also make sure that its actually tested since the code added in this PR is not.

Reply:
dont let me block this PR, we can just add a `TODO` in here tthat says something vague about sharing this logic with the stuff in search.cc in the future if that suits you for now?

Reply:
I added a TODO in code so we can consider changing this in the future.